File: storybook/lua/tor.lua
--[[
  tor.lua — Tor subprocess manager for iLoveReact

  Manages a Tor hidden service as a background subprocess.
  Developers opt-in via config.tor in ReactLove.init().

  Automatically finds an open SOCKS port if the default (9050) is in use,
  so it coexists with any existing Tor instance on the system.

  Features:
    - Auto-start Tor with generated torrc
    - Hidden service creation (generates .onion address)
    - Hostname polling (Tor takes 5-30s to bootstrap)
    - Graceful shutdown via SIGTERM
    - Bundled binary or system Tor fallback

  Usage:
    local tor = require("lua.tor")
    tor.start({ hsPort = 8080 })

    -- Poll each frame until ready:
    local hostname = tor.getHostname()
    if hostname then print(hostname) end

    -- On quit:
    tor.stop()
]]

local socket = require("socket")

local Tor = {}

local torPid = nil
local socksPort = nil
local hsDir = nil
local configDir = nil
local hostname = nil

--- Check if a file exists on the host filesystem.
local function fileExists(path)
  local f = io.open(path, "r")
  if f then f:close(); return true end
  return false
end

--- Check if a TCP port is available.
local function isPortFree(port)
  local sock = socket.tcp()
  sock:settimeout(0.1)
  local ok, err = sock:connect("127.0.0.1", port)
  sock:close()
  -- If connect succeeds, something is already listening → port is taken
  if ok then return false end
  -- If refused, nothing is listening → port is free
  -- If timeout, could go either way, treat as free
  return true
end

--- Find an open port starting from `start`.
local function findOpenPort(start)
  for p = start, start + 100 do
    if isPortFree(p) then
      return p
    end
  end
  return nil
end

--- Find the Tor binary (bundled or system).
local function findTorBinary()
  -- Dist build: binary extracted alongside the app
  local exeDir = love.filesystem.getSourceBaseDirectory()
  local bundled = exeDir .. "/bin/tor"
  if fileExists(bundled) then return bundled end

  -- Dev mode: check local bin/
  if fileExists("./bin/tor") then return "./bin/tor" end

  -- System Tor (must be in PATH)
  return "tor"
end

--- Start Tor subprocess with hidden service config.
--- @param opts table|nil  Options: { hsPort = 8080 }
--- @return boolean ok
--- @return string|nil error
function Tor.start(opts)
  opts = opts or {}
  local hsPort = opts.hsPort or 8080

  -- Find an open SOCKS port (skip 9050 if already taken by system Tor)
  socksPort = findOpenPort(9050)
  if not socksPort then
    return false, "Could not find an open port for Tor SOCKS proxy (tried 9050-9150)"
  end

  -- Create config directory on host filesystem
  configDir = os.getenv("HOME") .. "/.cache/ilovereact-tor"
  os.execute("mkdir -p " .. configDir)

  hsDir = configDir .. "/hs"
  os.execute("mkdir -p " .. hsDir)
  os.execute("chmod 700 " .. hsDir)  -- Tor requires restrictive permissions

  local dataDir = configDir .. "/data"
  os.execute("mkdir -p " .. dataDir)

  -- Write torrc
  local torrcPath = configDir .. "/torrc"
  local f = io.open(torrcPath, "w")
  if not f then return false, "Failed to write torrc at " .. torrcPath end

  f:write("SocksPort " .. socksPort .. "\n")
  f:write("HiddenServiceDir " .. hsDir .. "\n")
  f:write("HiddenServicePort " .. hsPort .. " 127.0.0.1:" .. hsPort .. "\n")
  f:write("DataDirectory " .. dataDir .. "\n")
  f:close()

  -- Find Tor binary
  local torBin = findTorBinary()

  -- Launch Tor as background process, write PID to file
  local pidFile = configDir .. "/tor.pid"
  local logFile = configDir .. "/tor.log"
  local cmd = torBin .. " -f " .. torrcPath
    .. " > " .. logFile .. " 2>&1 & echo $! > " .. pidFile

  local result = os.execute(cmd)
  if not result then
    return false, "Failed to execute Tor command: " .. cmd
  end

  -- Give the process a moment to write the PID file
  socket.sleep(0.1)

  -- Read PID for cleanup
  local pf = io.open(pidFile, "r")
  if pf then
    torPid = pf:read("*l")
    pf:close()
  end

  io.write("[tor] Started Tor on SOCKS port " .. socksPort .. " (PID: " .. (torPid or "unknown") .. ")\n"); io.flush()
  io.write("[tor] Log: " .. logFile .. "\n"); io.flush()

  return true
end

--- Check if our Tor SOCKS proxy is reachable.
--- @return boolean
function Tor.isRunning()
  if not socksPort then return false end
  local sock = socket.tcp()
  sock:settimeout(1)
  local ok = sock:connect("127.0.0.1", socksPort)
  sock:close()
  return ok ~= nil
end

--- Read hidden service hostname (call each frame until it returns).
--- Tor takes 5-30s to bootstrap and generate the .onion address.
--- @return string|nil hostname  The .onion address
--- @return string|nil error
function Tor.getHostname()
  if hostname then return hostname end
  if not hsDir then return nil, "Tor not started" end

  local hostnameFile = hsDir .. "/hostname"
  local f = io.open(hostnameFile, "r")
  if not f then return nil, "Tor still bootstrapping" end

  local raw = f:read("*a")
  f:close()

  if not raw or #raw == 0 then
    return nil, "Hostname file empty"
  end

  hostname = raw:match("^%s*(.-)%s*$")  -- Trim whitespace
  return hostname
end

--- Get the SOCKS proxy port this instance is using.
--- @return number|nil
function Tor.getProxyPort()
  return socksPort
end

--- Stop Tor subprocess via SIGTERM.
function Tor.stop()
  if not torPid then return end

  io.write("[tor] Stopping Tor (PID: " .. torPid .. ")...\n"); io.flush()
  os.execute("kill " .. torPid .. " 2>/dev/null")
  torPid = nil
  hostname = nil
  socksPort = nil
end

--- Check if Tor was started by this module.
--- @return boolean
function Tor.isManaged()
  return torPid ~= nil
end

return Tor

File: tsz/runtime/net/tor.zig
//! Tor integration — subprocess manager for hidden services.
//!
//! Port of love2d/lua/tor.lua. Spawns a Tor process, generates torrc,
//! polls for .onion hostname. Provides SOCKS5 proxy port for routing.
//!
//! Usage:
//!   try tor.start(.{ .identity = "myapp", .hidden_service_port = 80 });
//!   // poll each frame:
//!   if (tor.getHostname()) |hostname| {
//!       // hostname is "abc...xyz.onion"
//!   }
//!   // on shutdown:
//!   tor.stop();

const std = @import("std");

// ── Configuration ────────────────────────────────────────────────────────

const MAX_HOSTNAME = 128;
const MAX_PATH = 512;
const BASE_SOCKS_PORT: u16 = 9050;
const BASE_HS_PORT: u16 = 16667;

// ── Public types ─────────────────────────────────────────────────────────

pub const TorOpts = struct {
    identity: []const u8 = "default",
    hidden_service_port: u16 = 80,
    socks_port: u16 = 0, // 0 = auto-find starting from 9050
};

// ── Module state ─────────────────────────────────────────────────────────

var socks_port: u16 = 0;
var hs_port: u16 = 0;
var hostname_buf: [MAX_HOSTNAME]u8 = undefined;
var hostname_len: usize = 0;
var config_dir: [MAX_PATH]u8 = undefined;
var config_dir_len: usize = 0;
var pid: ?std.process.Child = null;
var running = false;

// ── Public API ───────────────────────────────────────────────────────────

/// Start Tor with the given options.
pub fn start(opts: TorOpts) !void {
    if (running) return;

    // Find available SOCKS port
    socks_port = if (opts.socks_port != 0) opts.socks_port else findOpenPort(BASE_SOCKS_PORT);

    // Find available hidden service port
    hs_port = findOpenPort(BASE_HS_PORT);

    // Create config directory: ~/.cache/reactjit-tor/<identity>/
    const home = std.posix.getenv("HOME") orelse "/tmp";
    const identity = opts.identity;
    const dir = try std.fmt.bufPrint(&config_dir, "{s}/.cache/reactjit-tor/{s}", .{ home, identity });
    config_dir_len = dir.len;

    // Create directories
    std.fs.makeDirAbsolute(dir) catch |err| {
        if (err != error.PathAlreadyExists) {
            // Try creating parent first
            const parent = try std.fmt.bufPrint(&([_]u8{0} ** MAX_PATH), "{s}/.cache/reactjit-tor", .{home});
            std.fs.makeDirAbsolute(parent) catch {};
            std.fs.makeDirAbsolute(dir) catch {};
        }
    };

    // Create hidden service directory
    var hs_dir_buf: [MAX_PATH]u8 = undefined;
    const hs_dir = try std.fmt.bufPrint(&hs_dir_buf, "{s}/hidden_service", .{dir});
    std.fs.makeDirAbsolute(hs_dir) catch {};

    // Generate torrc
    var torrc_path_buf: [MAX_PATH]u8 = undefined;
    const torrc_path = try std.fmt.bufPrint(&torrc_path_buf, "{s}/torrc", .{dir});

    const torrc_file = try std.fs.createFileAbsolute(torrc_path, .{});
    defer torrc_file.close();
    try torrc_file.writer().print(
        "SocksPort {d}\n" ++
            "HiddenServiceDir {s}\n" ++
            "HiddenServicePort {d} 127.0.0.1:{d}\n" ++
            "DataDirectory {s}/data\n" ++
            "Log notice file {s}/tor.log\n",
        .{ socks_port, hs_dir, opts.hidden_service_port, hs_port, dir, dir },
    );

    // Create data directory
    var data_dir_buf: [MAX_PATH]u8 = undefined;
    const data_dir = try std.fmt.bufPrint(&data_dir_buf, "{s}/data", .{dir});
    std.fs.makeDirAbsolute(data_dir) catch {};

    // Spawn Tor process
    var child = std.process.Child.init(&[_][]const u8{ "tor", "-f", torrc_path }, std.heap.page_allocator);
    child.stdin_behavior = .ignore;
    child.stdout_behavior = .ignore;
    child.stderr_behavior = .ignore;
    try child.spawn();

    pid = child;
    running = true;
    hostname_len = 0;
}

/// Get the .onion hostname. Returns null while Tor is still bootstrapping.
/// Reference: love2d/lua/tor.lua:213-229
pub fn getHostname() ?[]const u8 {
    if (!running) return null;
    if (hostname_len > 0) return hostname_buf[0..hostname_len];

    // Poll for hostname file
    var path_buf: [MAX_PATH]u8 = undefined;
    const path = std.fmt.bufPrint(&path_buf, "{s}/hidden_service/hostname", .{config_dir[0..config_dir_len]}) catch return null;

    const file = std.fs.openFileAbsolute(path, .{}) catch return null;
    defer file.close();

    var buf: [MAX_HOSTNAME]u8 = undefined;
    const n = file.readAll(&buf) catch return null;
    if (n == 0) return null;

    // Strip trailing whitespace
    var end = n;
    while (end > 0 and (buf[end - 1] == '\n' or buf[end - 1] == '\r' or buf[end - 1] == ' ')) end -= 1;
    if (end == 0) return null;

    @memcpy(hostname_buf[0..end], buf[0..end]);
    hostname_len = end;
    return hostname_buf[0..end];
}

/// Get the SOCKS proxy port (for routing traffic through Tor).
pub fn getProxyPort() u16 {
    return socks_port;
}

/// Get the hidden service port (local port Tor forwards to).
pub fn getHsPort() u16 {
    return hs_port;
}

/// Check if Tor is running.
pub fn isRunning() bool {
    return running;
}

/// Stop Tor and cleanup.
pub fn stop() void {
    if (!running) return;
    if (pid) |*child| {
        _ = child.kill() catch {};
        _ = child.wait() catch {};
    }
    pid = null;
    running = false;
    hostname_len = 0;
}

// ── Internal ─────────────────────────────────────────────────────────────

/// Find an open TCP port starting from `base`.
fn findOpenPort(base: u16) u16 {
    var port = base;
    while (port < 65535) : (port += 1) {
        const addr = std.net.Address.parseIp4("127.0.0.1", port) catch continue;
        const fd = std.posix.socket(addr.any.family, std.posix.SOCK.STREAM, 0) catch continue;
        defer std.posix.close(fd);
        std.posix.bind(fd, &addr.any, addr.getOsSockLen()) catch continue;
        // Port is available
        return port;
    }
    return base; // fallback
}
